prov/util/src/util_buf.c

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright (c) 2016 Intel Corporation. All rights reserved.
+ *
+ * This software is available to you under a choice of one of two
+ * licenses.  You may choose to be licensed under the terms of the GNU
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the
+ * BSD license below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+#include <fi_enosys.h>
+#include <fi_util.h>
+
+#define UTIL_BUF_ALIGN_SZ (16)
+
+static inline size_t util_buf_get_aligned_sz(size_t size)
+{
+	return ((size % UTIL_BUF_ALIGN_SZ) == 0) ?
+		size : ((size / UTIL_BUF_ALIGN_SZ) + 1) * UTIL_BUF_ALIGN_SZ;
+}
+
+struct util_buf_pool *util_buf_pool_create(size_t num, size_t size)
+{
+	int i, ret;
+	struct util_buf_pool *buf_pool;
+	struct util_buf *curr;
+	char *start_ptr;
+
+	buf_pool = (struct util_buf_pool *) calloc(1, sizeof(*buf_pool));
+	if (!buf_pool)
+		return NULL;
+
+	fastlock_init(&buf_pool->lock);
+	buf_pool->size = util_buf_get_aligned_sz(size);
+	buf_pool->num = num;
+	slist_init(&buf_pool->list);
+
+	ret = posix_memalign(&buf_pool->memory_region, UTIL_BUF_ALIGN_SZ,
+			     num * ((sizeof(struct util_buf) + size)));
+	if (ret) {
+		free(buf_pool);
+		return NULL;
+	}
+
+	start_ptr = (char *) buf_pool->memory_region;
+	for (i = 0; i < num; i++) {
+		curr = (struct util_buf *) (start_ptr +
+					    i * (size + sizeof(struct util_buf)));
+		curr->pool = buf_pool;
+		slist_insert_tail(&curr->entry, &buf_pool->list);
+	}
+	return buf_pool;
+}
```

Comment:
I think we want a more intelligent buffer pool as a general abstraction.  Rather than allocating all buffers up front, support a dynamically growing pool.  If size > 0, treat it as a max.  If size = 0, dynamically grow the pool.

The pool attributes may need to include a growth chunk size, plus buffer alignment